testing/src/Replay/WorkflowReplayer.php

```diff
@@ -4,20 +4,138 @@
 
 namespace Temporal\Testing\Replay;
 
-/** Replays a workflow given its history. Useful for backwards compatibility testing. */
+use RoadRunner\Temporal\DTO\V1\ReplayRequest;
+use RoadRunner\Temporal\DTO\V1\ReplayResponse;
+use Spiral\Goridge\Relay;
+use Spiral\Goridge\RPC\Codec\ProtobufCodec;
+use Spiral\Goridge\RPC\RPC;
+use Spiral\RoadRunner\Environment;
+use SplFileInfo;
+use Temporal\Api\Common\V1\WorkflowExecution;
+use Temporal\Api\Common\V1\WorkflowType;
+use Temporal\Testing\Replay\Exception\NonDeterministicWorkflowException;
+use Temporal\Testing\Replay\Exception\ReplayerException;
+use Temporal\Testing\Replay\Exception\RPCException;
+
+/**
+ * Replays a workflow given its history. Useful for backwards compatibility testing.
+ *
+ * @link https://docs.temporal.io/dev-guide/php/testing#replay
+ * @since RoadRunner 2023.3
+ */
 final class WorkflowReplayer
 {
+    private RPC $rpc;
+
+    public function __construct()
+    {
+        $this->rpc = new RPC(Relay::create(Environment::fromGlobals()->getRPCAddress()), new ProtobufCodec());
+    }
+
     /**
-     * Replays workflow from a resource that contains a json serialized history.
+     * Replays workflow from a history that will be loaded from Temporal server.
+     *
+     * @throws ReplayerException
      */
-    public function replayWorkflowExecutionFromFile(string $filename, string $workflowClass): void
-    {
-        $workflowExecutionHistory = WorkflowExecutionHistory::fromFile($filename);
-        $this->replayWorkflowExecution($workflowExecutionHistory, $workflowClass);
+    public function replayFromServer(
+        string $workflowType,
+        \Temporal\Workflow\WorkflowExecution $execution,
+    ): void {
+        $request = $this->buildRequest($workflowType, $execution);
+        $this->sendRequest('temporal.ReplayWorkflow', $request);
+    }
+
+    /**
+     * Downloads workflow history from Temporal server and saves it to a file.
+     *
+     * @param non-empty-string $workflowType
+     * @param non-empty-string $savePath
+     *
+     * @throws ReplayerException
+     */
+    public function downloadHistory(
```

Comment:
This should download the history and return it as a proto, I think. Then we should also provide a function for saving protos to a JSON file. This way the user can do everything in memory without touching disk, and enables inspecting history before replaying if they would like to (which they can't with `replayFromServer`). 

Semi-related: Where are the implementations of the other side of these `sendRequest` calls? I can't find the corresponding definitions.

Reply:
This is RR server side. https://github.com/temporalio/roadrunner-temporal/pull/394/files#diff-05a67dd647cb7c91ba1c5f7d5382094b4b839d86ca8d4e90e288f0566329c76cR188

Reply:
> Where are the implementations of the other side of these sendRequest calls? I can't find the corresponding definitions.

`$this->rpc->call($commad, $request);` does RoadRunner RPC call.
RFC about Replayer API there: https://github.com/roadrunner-server/roadrunner/issues/1640

Reply:
Ah, cool, I figured that was the case I just didn't realize I should go look in that other repo. Makes sense.

Reply:
Yes, that makes good sense.